### PR TITLE
[Python] new concept exercise - `basics`

### LINKED ISSUE #1999: [Python] Implement new Concept Exercise: basics
## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

The goal of this exercise is to teach the student the basics of programming in Python.

## Learning objectives

- Know what a variable (_name_)  is.
- Know how to name a variable.
- Know how to bind and re-bind a _name_ to a value (_assign or re-assign a value to a variable or assign a value to a constant_).
- Know what a `function` is and how  to define one via `def`.
- Know what the `return` keyword is, and how to use it to return a value from a function.
- Know how to call a function.
- Know how to declare an integer and a float.
- Know how to use mathematical operators `+` and `*` on integers and/or floats.
- Know how to define single- and multiline comments.
- Know how to define docstrings

## Out of scope

- built-in types
- built-in data structures
- Classes
- Methods and other function usages (_functions as arguments, nested functions_).
- Memory and performance characteristics.
- Default parameterrs
- Variadic parameters. (_*args and **kwargs_)

## Concepts

`basics`:  
- function definition & `return` keyword
- integers & floats, 
- mathematical operators `+` and `*`
- single and muliti-line comments.
- updating variable values 
- using functions
- variables, variable assignment, & variable naming

## Prerequisites

There are no prerequisites for this exercise

## Resources to refer to

[What is Python?](https://docs.python.org/3/faq/general.html#what-is-python)
[What is Python good for?](https://docs.python.org/3/faq/general.html#what-is-python-good-for)
[Numbers (_Python docs tutorial_)](https://docs.python.org/3/tutorial/introduction.html#numbers)
[Numeric Types](https://docs.python.org/3/library/stdtypes.html#numeric-types-int-float-complex)
[Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
[What is the difference between arguments and parameters in Python ?](https://docs.python.org/3/faq/programming.html#what-is-the-difference-between-arguments-and-parameters)

### Hints

See doc links above in `Resources to refer to`

### After

- Any concept exercise covering built-in data types or built-in functions.

## Representer

This exercise does not require any specific representation logic to be added to the representer.

## Analyzer

This exercise does not require any specific representation logic to be added to the analyzer..

## Implementing

Tests should be written using unittest.TestCase and the test file named lasanga_test.py.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

### PR DESCRIPTION
**Pull request by BethanyG**
May close #1999.

Added materials for new `basics` concept exercise in Python.

### CODE REVIEW

**Comment by BethanyG** [comment]
I've added two things to this that may be a bit of debate:  comments and docstrings.  Both felt as though they really didn't need their own exercise (_I know - we have an open PR...nevertheless_).  

Additionally, I omitted specific examples of classes since Python doesn't "require" them as a programming construct.  I also omitted calling methods in the instructions, but added some language in the `after.md` for them.  It does make the language "fuzzy", and the `after.md` a little complicated...but I didn't want to get too deep into weird details.

File: languages/python/exercises/concept/basics/.docs/after.md [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+>>> start_text = "my silly sentence for examples."
+>>> str.upper(start_text)
+"MY SILLY SENTENCE FOR EXAMPLES."
+
+#this is a method call of an *instance* of the str *class*
+>>> start_text.upper()
+"MY SILLY SENTENCE FOR EXAMPLES."
+```
+
+[Comments](https://realpython.com/python-comments-guide/#python-commenting-basics) in Python start with a `#` (_hash character_) that is not part of a string, and end at line termination. Unlike many other programming languages, Python does not support multi-line comment marks. Each line of a comment block must start with the `#` character. Comments are ignored by the interpreter:
```
Comment (iHiD) [inline languages/python/exercises/concept/basics/.docs/after.md:105]: I think leave the "hash character" bit out as people have different names for this.

```suggestion
[Comments](https://realpython.com/python-comments-guide/#python-commenting-basics) in Python start with a `#` that is not part of a string, and end at line termination. Unlike many other programming languages, Python does not support multi-line comment marks. Each line of a comment block must start with the `#` character. Comments are ignored by the interpreter:
```

File: languages/python/exercises/concept/basics/.docs/introduction.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+MY_FIRST_CONSTANT = 16
+
+# Re-assignment will be allowed by the compiler & interpreter,
+# but is VERY strongly discouraged.
+# Please don't do: MY_FIRST_CONSTANT = "Some other value"
+```
+
+In Python, units of functionality are encapsulated in [functions](https://docs.python.org/3/reference/compound_stmts.html#function). When functions are bound to a [class](https://docs.python.org/3/reference/datamodel.html#classes) name, they're referred to as [methods](https://docs.python.org/3/c-api/method.html#method-objects).
+
+The keyword `def` begins a [function definition](https://docs.python.org/3/tutorial/controlflow.html#defining-functions). It must be followed by the function name and a parenthesized list of zero or more formal [parameters](https://docs.python.org/3/glossary.html#term-parameter). The `def` line is terminated with a colon. Statements for the _body_ of the function begin on the next line, and must be _indented in a block_. Functions explicitly return a value or object via the `return` keyword:
```
Comment (ErikSchierboom) [inline languages/python/exercises/concept/basics/.docs/introduction.md:27]: Python is white-space sensitive, right? I don't know any Python, but I was wondering if it would be helpful to explain how to indent properly (can I use tabs/spaces?, how many tabs/spaces?) as I feel that is something that might trip up new students.

File: languages/python/exercises/concept/basics/.docs/after.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+Python puts a strong emphasis on code readability and (_similar to Haskell_) uses significant indentation to denote function, method, and class definitions. The [zen of Python (PEP 20)](https://www.python.org/dev/peps/pep-0020/) lays out additional inspiration and philosophy.
+
+Objects are [assigned](https://docs.python.org/3/reference/simple_stmts.html#assignment-statements) to [names](https://docs.python.org/3/reference/executionmodel.html#naming-and-binding) in Python via the `=` or _assignment operator_. [Variables](https://realpython.com/python-variables/) are written in [`snake_case`](https://en.wikipedia.org/wiki/Snake_case), and _constants_ usually in `SCREAMING_SNAKE_CASE`. A name (_variable or constant_) is not itself _typed_, and can be attached or re-attached to different objects over its lifetime. For extended naming conventions and advice, see [PEP 8](https://www.python.org/dev/peps/pep-0008/).
+
+```python
+>>> my_first_variable = 1
+>>> my_first_variable = {"bear" : "brown", "cat": "black"}
```
Comment (ErikSchierboom) [inline languages/python/exercises/concept/basics/.docs/after.md:11]: Maybe remove this line to not possibly confuse the student with unfamiliar syntax?

```suggestion
```

File: languages/python/exercises/concept/basics/.docs/hints.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+## 1. Define expected bake time in minutes
+
+- You need to [name](https://realpython.com/python-variables/) a constant, and [assign](https://docs.python.org/3/reference/simple_stmts.html#grammar-token-assignment-stmt) it an integer value.
+
+## 2. Calculate remaining bake time in minutes
+
+- You need to define a [function](https://docs.python.org/3/tutorial/controlflow.html#defining-functions) with a single parameter representing the time elapsed so far.
+- Use the [mathematical operator for subtraction](https://docs.python.org/3/tutorial/introduction.html#numbers) to subtract values.
+- This function should `return` a value.
```
[the remark this answers is not in the document]
Comment (cmccandless) [reply languages/python/exercises/concept/basics/.docs/hints.md:14]: @ErikSchierboom your suggested change contains an extra closing bracket.

File: languages/python/exercises/concept/basics/.docs/introduction.md (hunk 2 above)
Comment (cmccandless) [reply languages/python/exercises/concept/basics/.docs/introduction.md:27]: > Python is white-space sensitive, right? 

Correct.

> can I use tabs/spaces?, how many tabs/spaces?

I couldn't find a more recent version of this doc, but it's still correct: https://docs.python.org/2.0/ref/indentation.html

TLDR: You can use either, but be consistent and readable. Mixing tabs and spaces in the same file is a `SyntaxError`

@ErikSchierboom On a related note: has any other track done an exercise on linters and formatters? It's quite common to at least use `flake8` or `pylint` to check your Python code for style violations and `SyntaxError`s, and there are a few good formatters out there as well (`black` and `autopep8` to name a couple).

Comment (ErikSchierboom) [reply languages/python/exercises/concept/basics/.docs/introduction.md:27]: @cmccandless We haven't, because linters and formatters are usually not part of the language. I think having an exercise for them makes less sense, but it would definitely be something that we can include in instructions for CLI users (we have a [special file for this in v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#exercisesshareddocsclimd), which will automatically apply to all exercises in a track).

File: languages/python/exercises/concept/basics/.docs/hints.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply languages/python/exercises/concept/basics/.docs/hints.md:14]: Whoops, great catch!

File: languages/python/exercises/concept/basics/.docs/instructions.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+## 4. Calculate total elapsed cooking time (prep + bake) in minutes
+
+Implement the `elapsed_time_in_minutes()` function that has two parameters: `number_of_layers` (_the number of layers added to the lasagna_) and `elapsed_bake_time` (_the number of minutes the lasagna has been baking in the oven_). This function should return the total number of minutes you've been cooking, or the sum of your preparation time and the time the lasagna has already spent baking in the oven.
+
+```python
+>>> elapsed_time_in_minutes(3, 20)
+26
+```
+
+## 5. Add docstrings and comments to your work
```
[the remark this answers is not in the document]
Comment (BethanyG) [reply languages/python/exercises/concept/basics/.docs/instructions.md:41]: Hard agree.  :-)  I was debating it when I wrote it because adding "notes" might lead someone to put in things about what page the recipe was on, etc....but I think we have enough hints and links to let the student know it's for their code -- and it will read better.

File: languages/python/exercises/concept/basics/.docs/introduction.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+str(object='') -> str
+str(bytes_or_buffer[, encoding[, errors]]) -> str
+
+Create a new string object from the given object. If encoding or
+errors is specified, then the object must expose a data buffer
+that will be decoded using the given encoding and error handler.
+Otherwise, returns the result of object.__str__() (if defined)
+or repr(object).
+encoding defaults to sys.getdefaultencoding().
+errors defaults to 'strict'.
```
[the remark this answers is not in the document]
Comment (BethanyG) [reply languages/python/exercises/concept/basics/.docs/introduction.md:94]: Oops.  I thought I'd taken that out of the `introduction.md`.  I meant that extra one for the `after.md`.  I'll remove it.

File: languages/python/exercises/concept/basics/.docs/hints.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (BethanyG) [reply languages/python/exercises/concept/basics/.docs/hints.md:14]: Changed to 
> This function should [return a value](https://docs.python.org/3/reference/simple_stmts.html#return).

File: languages/python/exercises/concept/basics/.docs/introduction.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
+Python is a [dynamic and strongly](https://stackoverflow.com/questions/11328920/is-python-strongly-typed) typed [object-oriented](https://en.wikipedia.org/wiki/Object-oriented_programming) programming language. It supports both imperative (_object-oriented, procedural_) and declarative (_functional, concurrent_) programming paradigms. Python emphasizes code readability and (similar to Haskell) uses significant indentation.
```
[the remark this answers is not in the document]
Comment (BethanyG) [reply languages/python/exercises/concept/basics/.docs/introduction.md:1]: Removed everything expect the ref to `snake_case` - since that is a weird-ish name.

File: languages/python/exercises/concept/basics/.docs/introduction.md (hunk 2 above)
Comment (BethanyG) [reply languages/python/exercises/concept/basics/.docs/introduction.md:25]: **PENDANT ALERT**  😉 

> Python is white-space sensitive, right?

The reason I used "significant indent" is because _some_ whitespace matters in Python ..but other whitespace doesn't -- or is implicitly concatenated by the interpreter.  The [lexical analysis docs](https://docs.python.org/3/reference/lexical_analysis.html#line-structure) draw distinctions between _logical lines_ and _physical lines_.  It can be a bit annoying.  For example, whitespace in a dictionary literal, tuple, list, or comprehension is implicitly concatenated.  But if the declaration appears inside a function or class and you _don't_ indent the logical line, then you get an `IndentationError`: 

```python
#this is implicitly concatenated by the interpreter
indented_dict = {
                 "cat" : "orange tabby", 
                 "dog": "brown beagle",
                 "bird": "blue mockingjay"
                }

#this is too
indented_dict_two = {"cat" : "orange tabby", 
                     "dog": "brown beagle",
                     "bird": "blue mockingjay"}

#this is perfectly fine, and won't toss any errors.
one_line_dict = {"cat" : "orange tabby", "dog": "brown beagle", "bird": "blue mockingjay"}

#but this will toss an error, because the dict variable name isn't indented properly.
def example_function():
    example_variable = "this is a string."
  indented_dict = {
                 "cat" : "orange tabby", 
                 "dog": "brown beagle",
                 "bird": "blue mockingjay"
                }
```

Pedantry aside, I added this entry on [indentation](https://docs.python.org/3/reference/lexical_analysis.html#indentation) to the `after.md` (_looks like Python 3.8 now has a `TabError`...a subclass of `IndentationError`.  lol_), and added the following to the `introduction.md`:

>  Statements for the _body_ of the function begin on the next line, and must be _indented in a block_. There is no strict indentation amount (_either space **OR** [tab] characters are acceptable_), but indentation must be _consistent for all indented lines_.

File: languages/python/exercises/concept/basics/.docs/introduction.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+#this is a single line comment
+
+x = "foo"  #this is an in-line comment
+
+#this is a multi-line
+#comment block over multiple lines
+#these should be used sparingly
+```
+
+The first statement of a function body can optionally be a [docstring](https://docs.python.org/3/tutorial/controlflow.html#tut-docstrings), which concisely summarizes the function or object's purpose. These docstrings are read by automated documentation tools, and are returned by calling `__doc__` on the function, method, or class. They are recommended for programs of any size where documentation is needed:
```
[the remark this answers is not in the document]
Comment (BethanyG) [reply languages/python/exercises/concept/basics/.docs/introduction.md]: Removed in update.

**Comment by BethanyG** [comment]
Going to leave this here just a little longer, in case I've missed something.

File: languages/python/exercises/concept/basics/.docs/instructions.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+## 4. Calculate total elapsed cooking time (prep + bake) in minutes
+
+Implement the `elapsed_time_in_minutes()` function that has two parameters: `number_of_layers` (_the number of layers added to the lasagna_) and `elapsed_bake_time` (_the number of minutes the lasagna has been baking in the oven_). This function should return the total number of minutes you've been cooking, or the sum of your preparation time and the time the lasagna has already spent baking in the oven.
+
+```python
+>>> elapsed_time_in_minutes(3, 20)
+26
+```
+
+## 5. Update the recipe with notes
```
Comment (ErikSchierboom) [inline languages/python/exercises/concept/basics/.docs/instructions.md:41]: I like this updated task description!

**Comment by iHiD** [comment]
Whoop!! Great work :)

**Comment by ErikSchierboom** [comment]
Hurray! We have a Python basics exercise 🦜
